apps/src/blockly/addons/blockSvgLimitIndicator.ts

```diff
@@ -0,0 +1,131 @@
+import {BlockSvg} from 'blockly';
+
+/**
+ * Represents a bubble on a Blockly block, displaying the count of blocks remaining
+ * based on a limit initially stated in the toolbox XML.
+ */
+export default class BlockSvgLimitIndicator {
+  private readonly blockSvg: BlockSvg;
+  private count: number;
+  private readonly bubbleSize: number = 18;
+  private readonly halfBubbleSize: number;
+  private limitGroup: SVGElement | undefined;
+  private limitRect: SVGElement | undefined;
+  private limitText: SVGElement | undefined;
+
+  /**
+   * Constructs an SVG group to track a block limit.
+   * @param {Element} element - The block associated with the limit.
+   * @param {string} [count] - The initial count to display.
+   */
+  constructor(element: BlockSvg, count: number) {
+    this.blockSvg = element;
+    this.count = count;
+    this.halfBubbleSize = this.bubbleSize / 2;
+
+    this.limitGroup = undefined;
+    this.limitRect = undefined;
+    this.limitText = undefined;
+    this.initializeSvgElements();
+  }
+
+  /**
+   * Updates the displayed count within the limit bubble and adjusts styling based
+   * on the count.
+   * @param {number} newCount The new count to display.
+   */
+  public updateCount(newCount: number) {
+    this.count = newCount;
+    this.updateTextAndClass();
+  }
+
+  /**
+   * Updates the text and class of the limit bubble depending on whether the student
+   * has blocks remaining. If over the limit, displays an exclamation mark and changes
+   * the bubble and text color.
+   */
+  private updateTextAndClass() {
+    if (!this.limitText || !this.limitGroup) {
+      return;
+    }
+    if (this.count >= 0) {
+      this.limitText.textContent = `${this.count}`;
+      Blockly.utils.dom.removeClass(this.limitGroup, 'overLimit');
+      Blockly.utils.dom.removeClass(this.limitText, 'overLimit');
+    } else {
+      this.limitText.textContent = '!';
+      Blockly.utils.dom.addClass(this.limitGroup, 'overLimit');
+      Blockly.utils.dom.addClass(this.limitText, 'overLimit');
+    }
+    this.render();
+  }
+
+  /**
+   * Initializes the SVG elements that make up the limit bubble.
+   */
+  private initializeSvgElements() {
+    this.limitGroup = Blockly.utils.dom.createSvgElement(
```

Comment:
We could avoid having to do a bunch of null checks on `limitGroup` and `limitRect` by assigning them in the constructor rather than a helper method. Then we would know they are never actually undefined.

Reply:
Awesome idea!